# Independent Lab: Advanced Data Manipulation
Michelle Calderwood | BGEN 632 – Spring 2025

This notebook demonstrates advanced data wrangling techniques using pandas and piping notation. We will rename columns, filter/select with regex, and use chaining to clean and explore data efficiently.

### Import libraries

In [104]:
import os
import pandas as pd

# Set working directory (if needed)
os.chdir("/Users/michellecalderwood/Documents/GitHub/week8labs/data")
print(os.getcwd())

# Load dataset
turnover_df = pd.read_csv("gh_turnover_four_projects.csv")
turnover_df.info()

/Users/michellecalderwood/Documents/GitHub/week8labs/data
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memo

### Rename Columns
Rename columns to more descriptive names to enhance clarity and consistency.

In [106]:
turnover_df = turnover_df.rename(columns={
    'domain': 'project_domain',
    'language': 'project_language',
    'windows': 'project_age',
    'window_idx': 'quarter',
    'num_team': 'team_size',
    'num_commits': 'project_commits',
    'blau_gender': 'gender_blau',
    'Gini_gh_ten': 'gh_tenure_gini',
    'Core1': 'core_dev',
    'commits': 'user_commits',
    'propCommits': 'commit_proportion',
    'github_tenure': 'gh_tenure',
    'leavesNextQ': 'leaves_next_q'
})

turnover_df.columns

Index(['project_id', 'project_domain', 'project_language', 'project_age',
       'quarter', 'team_size', 'project_commits', 'gender_blau',
       'gh_tenure_gini', 'user_id', 'core_dev', 'user_commits',
       'commit_proportion', 'gender', 'gh_tenure', 'leaves_next_q'],
      dtype='object')

### Show Mean and Median of user_commits using pandas piping

In [112]:
(
 turnover_df
 .agg({"user_commits": ["mean", "median"]})
)

user_commits
mean      339.365751
median     18.000000

### Select and Filter Combined (mimics tidyverse)
Select columns with p or g, filter project_commits > 9000

In [118]:
turnover_newer = (
 turnover_df
 .filter(regex='(^p)|(g)')
 .query("project_commits > 9000")
)

turnover_newer.head()

project_id project_domain project_language  project_age  project_commits  \
0        1101          OTHER           Python           23            22688   
1        1101          OTHER           Python           23            22688   
2        1101          OTHER           Python           23            22688   
3        1101          OTHER           Python           23            22688   
4        1101          OTHER           Python           23            22688   

   gender_blau  gh_tenure_gini gender  gh_tenure  
0          0.0        0.422259   male        426  
1          0.0        0.422259   male         46  
2          0.0        0.422259   male        941  
3          0.0        0.422259   male       1870  
4          0.0        0.422259   male        781

### Switch to Original DataFrame
Filter columns ending in 'e' or containing 'q', then gh_tenure >= 1991

In [120]:
subset_df = (
 turnover_df
 .filter(regex='(e$)|(q)')
 .query("gh_tenure >= 1991")
)

subset_df.head()

project_language  project_age  quarter  team_size  gh_tenure  \
339             Java           16       17          2       2006   
457           Python           23       24         20       2054   

     leaves_next_q  
339              0  
457              0

### Mean and Median of user_commits in subset
Update filter like this so user_commits is kept

In [122]:
subset_df = (
    turnover_df
    .filter(regex='(e$)|(q)|(^user_commits$)')  # include user_commits explicitly
    .query("gh_tenure >= 1991")
)
(
    subset_df
    .agg({"user_commits": ["mean", "median"]})
)

user_commits
mean             3.5
median           3.5

### Summary
Mean and Median of `user_commits`: Original vs Filtered Subset

We used pandas piping to select:

- Columns that end in the letter `e` or contain the letter `q`
- Rows where `gh_tenure >= 1991`
- We also made sure to include `user_commits` manually in our column filter

#### 📊 Results:

| Statistic     | Full Dataset | Filtered Subset (gh_tenure ≥ 1991) |
|---------------|--------------|------------------------------------|
| Mean          | 339.37       | 3.50                               |
| Median        | 18.00        | 3.50                               |

#### 💡 Interpretation:

The mean and median of `user_commits` drop significantly in the filtered subset. This suggests that contributors who remained active longer (`gh_tenure >= 1991`) were **less active in terms of commit counts** than the average contributor. This may imply:

- A core group of long-term contributors contributed minimally
- Or that high-commit contributors had shorter tenures

This contrast highlights the importance of exploring both tenure and contribution volume in turnover and participation studies.

### Reference

[ChatGPT_Independent.ipynb](ChatGPT_Independent.ipynb)